test/extended/oauth/htpasswd.go

```diff
@@ -0,0 +1,90 @@
+package oauth
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+var _ = g.Describe("[Serial][Suite:openshift/oauth/serial] Htpasswd identity provider", func() {
+	defer g.GinkgoRecover()
+	var (
+		oc                    = exutil.NewCLI("htpasswd-configure", exutil.KubeConfigPath())
+		testDataBaseDir       = exutil.FixturePath("testdata", "oauth_idp")
+		htpasswdFixture       = filepath.Join(testDataBaseDir, "oauth-htpasswd.yaml")
+		htpasswdSecretFixture = filepath.Join(testDataBaseDir, "htpasswd-secret.yaml")
+		oauthNoIdpFixture     = filepath.Join(testDataBaseDir, "oauth-noidp.yaml")
+		authns                = "openshift-authentication"
+	)
+	g.Context("", func() {
+		g.AfterEach(func() {
+			err := oc.AsAdmin().KubeClient().CoreV1().Secrets("openshift-config").Delete("htpasswd", &metav1.DeleteOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+			e2e.Logf("now resetting oauth cluster")
+			e2e.Logf("removing htpasswd idp from oauth")
+			err = oc.AsAdmin().Run("replace").Args("-f", oauthNoIdpFixture).Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+			err = wait.Poll(5*time.Second, 5*time.Minute, func() (done bool, err error) {
+				pods, err := oc.AsAdmin().KubeClient().CoreV1().Pods(authns).List(metav1.ListOptions{})
+				if err != nil {
+					e2e.Logf("could not list pods in %s namespace", authns)
+					return false, nil
+				}
+				// wait for all pods running in openshift-authentication ns
+				for _, pod := range pods.Items {
+					if pod.Status.Phase != v1.PodRunning {
+						e2e.Logf("pod %s in %s not running yet", pod.Name, authns)
+						return false, nil
+					}
+				}
+				return true, nil
+			})
+			o.Expect(err).NotTo(o.HaveOccurred())
+			// cleanup created users/identity
+			err = oc.AsAdmin().Run("delete").Args("identities", "htpasswd:testuser").Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+			err = oc.AsAdmin().Run("delete").Args("user", "testuser").Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+		})
+
+		g.It("should successfully be configured", func() {
+			e2e.Logf("configuring htpasswd")
+			patch, err := ioutil.ReadFile(htpasswdFixture)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			oauthPatch := string(patch)
+			oc.SetNamespace("openshift-config")
+			err = oc.AsAdmin().Run("apply").Args("-f", htpasswdSecretFixture).Execute()
+			err = oc.AsAdmin().Run("patch").Args("oauth", "cluster", "-p", oauthPatch, "--type=merge").Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+			e2e.Logf("logging in as htpasswd user testuser")
+			err = wait.Poll(10*time.Second, 5*time.Minute, func() (done bool, err error) {
+				out, err := oc.Run("login").Args("-u", "testuser").InputString("password" + "\n").Output()
```

Comment:
You shouldn't retry logins, instead wait for the operator to stop reporting `Progressing=True`, login should work right after that. Perhaps fail if `Degraded=True` or `Available=True` once `Progrressing=False`?

I'd suggest have a separate function that checks the above, it will come in handy with other IdPs as well.

Reply:
good point, yes, added the CO Progressing=false/Available=True/Degraded=false check.  Still need the retry on login, (or a few second pause?)